### PR TITLE
✨ [bento] Preact version of amp-sidebar

### PR DESCRIPTION
**Pull request by krdwan**
First draft of Preact amp-sidebar component for initial review

12/11 - Update
1. Added unit tests - 
- imperative API
- click to open, close, toggle
- click mask to close 
- left and right sides
- animations
2. Updated storybook
- removed background content
- add buttons to open, close, toggle to within sidebar
3. Added type definition file

### CODE REVIEW

File: extensions/amp-sidebar/1.0/sidebar.js [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+  useImperativeHandle,
+  useRef,
+  useState,
+} from '../../../src/preact';
+import {useStyles} from './sidebar.jss';
+
+const ANIMATION_DURATION = 350;
+const ANIMATION_PRESETS = {
+  fadeIn: [{opacity: 0}, {opacity: 1}],
+  slideInLeft: [{transform: 'translateX(-100%)'}, {transform: 'translateX(0)'}],
```
Comment (dvoytenko) [inline extensions/amp-sidebar/1.0/sidebar.js:33]: These names, most likely, should be in quotes. I.e. this is not an enum, but rather a simple string-keyed object map. Especially since you use string keys below to access this map. However, I'd also point out: do we really need a map here? Why not just have three plain constants?

File: extensions/amp-sidebar/1.0/sidebar.js [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+/**
+ * @param {!AccordionDef.Props} props //TODO: will update
+ * @param ref
+ * @return {PreactDef.Renderable}
+ */
+function SidebarWithRef(
+  {
+    as: Comp = 'div',
+    side = 'left',
```
Comment (dvoytenko) [inline extensions/amp-sidebar/1.0/sidebar.js:58]: Possible ok for the first PR, but does the previous versions of sidebar refer to this as left|right? Or something more LTR/RTL neutral?

File: extensions/amp-sidebar/1.0/sidebar.js [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+      close,
+      toggle,
+    }),
+    [open, close, toggle]
+  );
+
+  useEffect(() => {
+    const sidebarElement = sidebarRef.current;
+    const maskElement = maskRef.current;
+    if (sidebarElement == undefined || maskElement == undefined) {
```
Comment (dvoytenko) [inline extensions/amp-sidebar/1.0/sidebar.js:110]: `== null` for both.

File: extensions/amp-sidebar/1.0/sidebar.js [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    }
+
+    let sidebarAnimation;
+    let maskAnimation;
+    // "Make Visible" Animation
+    if (opened) {
+      const postVisibleAnim = () => {
+        setStyle(sidebarElement, 'transform', 'translateX(0)');
+        setStyle(maskElement, 'opacity', 1);
+        sidebarElement./*REVIEW*/ focus();
```
Comment (dvoytenko) [inline extensions/amp-sidebar/1.0/sidebar.js:121]: Does this do anything at all? Is the intention to focus the first focusable element in the sidebar? That'd be a good idea, of course, but I'm not sure this does that.

File: extensions/amp-sidebar/1.0/sidebar.js [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+          }`}
+          role="menu"
+          tabindex="-1"
+          {...rest}
+        >
+          {children}
+        </ContainWrapper>
+        <div
+          ref={(r) => {
+            maskRef.current = r;
```
Comment (dvoytenko) [inline extensions/amp-sidebar/1.0/sidebar.js:204]: Just `ref={maskRef}`?

File: extensions/amp-sidebar/1.0/sidebar.jss.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {createUseStyles} from 'react-jss';
+
+const baseClass = {
```
Comment (dvoytenko) [inline extensions/amp-sidebar/1.0/sidebar.jss.js:19]: The meaning of "baseClass" is not quite clear.

File: extensions/amp-sidebar/1.0/sidebar.jss.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  minWidth: '45px !important',
+  outline: 'none',
+  overflowX: 'hidden !important',
+  overflowY: 'auto !important',
+  zIndex: 2147483647,
+};
+
+const left = {
+  left: 0,
+  transform: 'translateX(-100%)',
```
Comment (dvoytenko) [inline extensions/amp-sidebar/1.0/sidebar.jss.js:35]: Can we avoid `transform` here and below? It's best to define the stylesheets with the final values and leave offset values to the animation.

File: extensions/amp-sidebar/1.0/sidebar.jss.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+  position: 'fixed !important',
+  top: '0 !important',
+  left: '0 !important',
+  width: '100vw !important',
+  height: '100vh !important',
+  /* Prevent someone from making this a full-sceen image */
+  backgroundImage: 'none !important',
+  backgroundColor: 'rgba(0, 0, 0, 0.5)',
+  zIndex: 2147483646,
+  opacity: 0,
```
Comment (dvoytenko) [inline extensions/amp-sidebar/1.0/sidebar.jss.js:53]: Ditto. It's best to have "final" values and leave animation state to animations.

**Review by dvoytenko — COMMENTED** [review]
This looks pretty much there. No significant comments on structure.

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 2 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.js:58]: Good point.  Yes the previous version of sidebar refers to this as "left" and "right".

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 5 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.js:204]: Thanks updated, I was meaning to ask, is there a reason the `ContainWrapper` above uses this type of syntax?

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 3 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.js:110]: Updated in latest!

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 1 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.js:33]: Good point, we can just use three constants for this.  Thanks!

File: extensions/amp-sidebar/1.0/sidebar.jss.js (hunk 7 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.jss.js:35]: This is a good point and I also considered this.  What I noticed is that since the order is Render with original css (from jss here).  Then do the animation.  I noticed that if we remove this line, Preact will first render the component for an instant, then start the animation.  So in this case, it will flash the sidebar fully opened, then start animating it from the side.  Same with opacity below and the mask.

File: extensions/amp-sidebar/1.0/sidebar.jss.js (hunk 8 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.jss.js:53]: Ditto: as above, it will flash the mask for an instant before animating it in if we remove this line.

File: extensions/amp-sidebar/1.0/sidebar.jss.js (hunk 6 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.jss.js:19]: Thanks, updated to "sidebarClass".

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 3 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.js:110]: As a precedent for consistent styling, I also considered

`!sidebarElement || !maskElement`.  What do you think?

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 4 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.js:121]: I will remove for now and address in a future PR.

In the previous sidebar, the focus is shifted to an invisible close button.

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 3 above)
Comment (dvoytenko) [reply extensions/amp-sidebar/1.0/sidebar.js]: Yes. That's better.

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 4 above)
Comment (dvoytenko) [reply extensions/amp-sidebar/1.0/sidebar.js]: That's actually a good idea. But if it's coming in the follow up PRs, feel free to defer.

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 5 above)
Comment (dvoytenko) [reply extensions/amp-sidebar/1.0/sidebar.js]: I don't see any reason, tbh.

File: extensions/amp-sidebar/1.0/sidebar.jss.js (hunk 7 above)
Comment (dvoytenko) [reply extensions/amp-sidebar/1.0/sidebar.jss.js:35]: Sure, that definitely should be taken care of. There are a couple of solutions here. But I think in this case the easiest for us is:

1. Switch to `useLayoutEffect()`
2. Set the initial pre-animation state before animation is started.

E.g.

```
useLayoutEffect(() => {
  ...
  if (opened) {
    ...
    ...
    setStyles(sidebarElement, SLIDE_FROM_LEFT_INI_STYLES);
    sidebarAnimation = sidebarElement.animate(SLIDE_FROM_LEFT_ANIM);
  }
}, ...);
```

Typically none of this is need in the closing animations - only opening has this "initialization" problem.

Why this is better? Because right in that scope you're also setting final styles, such as `setStyle(sidebarElement, 'transform', 'translateX(0)');`. Thus it makes it very clear which styles are affected by the animation from start to end.

And, btw, if you set the ini style before animation to something like `transform: translateX(-100%)`, that would mean that the best thing is to reset it to empty string in the end instead of `translateX(0)` (which is equivalent to removing it from style). That's because the animation effect is temporary. And it's nice that it'd set itself while running and remove itself cleanly once done.

File: extensions/amp-sidebar/1.0/sidebar.jss.js (hunk 8 above)
Comment (dvoytenko) [reply extensions/amp-sidebar/1.0/sidebar.jss.js:53]: Deduping this discussion to [this thread](https://github.com/ampproject/amphtml/pull/31479#discussion_r537843458).

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 4 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.js]: Thanks, added to the tracker item: https://github.com/ampproject/amphtml/issues/31366

File: extensions/amp-sidebar/1.0/sidebar.jss.js (hunk 7 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.jss.js]: Awesome, thanks for the thorough explanation.  I've updated in the latest revision and factored out the styles as you've given in the example so that they are module level constants.  If you have opinions on the re-factoring please let me know, it might be overkill since we only apply two styles in animation.

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 5 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.js]: Updated to `ref={sidebarRef}` in the lines above as well

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 3 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.js]: Updated in the latest!

File: extensions/amp-sidebar/1.0/sidebar.js [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+function useValueRef(current) {
+  const valueRef = useRef(null);
+  valueRef.current = current;
+  return valueRef;
+}
+/**
+ * @param {!Element} element
+ * @param {!Object<string, *>} styles
+ */
+function safelySetStyles(element, styles) {
```
Comment (dvoytenko) [inline extensions/amp-sidebar/1.0/sidebar.js:63]: Why is this needed?

File: extensions/amp-sidebar/1.0/sidebar.js [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+          role="menu"
+          tabindex="-1"
+          {...rest}
+        >
+          {children}
+        </ContainWrapper>
+        <div
+          ref={maskRef}
+          onClick={() => close()}
+          className={`${classes.maskClass}`}
```
Comment (dvoytenko) [inline extensions/amp-sidebar/1.0/sidebar.js:231]: Just `className={classes.maskClass}`

File: extensions/amp-sidebar/1.0/sidebar.js [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+const ANIMATION_KEYFRAMES_SLIDE_IN_RIGHT = [
+  {'transform': 'translateX(100%)'},
+  {'transform': 'translateX(0)'},
+];
+
+const ANIMATION_STYLES_SIDEBAR_LEFT_INIT = {'transform': 'translateX(-100%)'};
+const ANIMATION_STYLES_SIDEBAR_RIGHT_INIT = {'transform': 'translateX(100%)'};
+const ANIMATION_STYLES_MASK_INIT = {'opacity': '0'};
+const ANIMATION_STYLES_SIDEBAR_FINAL = {'transform': ''};
+const ANIMATION_STYLES_MASK_FINAL = {'opacity': '1'};
```
Comment (dvoytenko) [inline extensions/amp-sidebar/1.0/sidebar.js:47]: Should be enough to do `{'opacity': ''}` (empty string).

**Review by dvoytenko — CHANGES_REQUESTED** [review]
Looks good, but one thing is not quite clear.

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 9 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.js:63]: So the thought process is like this:

In order to have styles defined in the module level constants, and not directly inline in `setStyles` then we have to include the `assertDoesNotContainDisplay` wrapper around the styles object that is passed in.  i.e. 
```
setStyles(element, {opacity: 1}) //okay

const style = {opacity:1}
setStyles(element, style) //not okay
setStyles(element, assertDoesNotContainDisplay(style)) //okay
```

So I thought it made sense to include the wrapper function `safelySetStyles` to clean up the syntax where ever we use `setStyles`

What do you think?  We can also put the `assertDoesNotContainDisplay` directly inline, but it looks a lot less clean in my opinion.

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 11 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.js]: Thanks good catch!! 😅😅

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 10 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.js]: Thanks, updated!  😅😅

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 9 above)
Comment (dvoytenko) [reply extensions/amp-sidebar/1.0/sidebar.js:63]: Ok. I see. I suppose this is ok then. Obviously seems weird since the styles are right there. But I suppose if tooling forces us into this then it's ok.

**Review by dvoytenko — APPROVED** [review]

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 9 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.js:63]: Sounds good!

**Review by dvoytenko — APPROVED** [review]

File: extensions/amp-sidebar/1.0/sidebar.js [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+const ANIMATION_STYLES_MASK_INIT = {'opacity': '0'};
+const ANIMATION_STYLES_SIDEBAR_FINAL = {'transform': ''};
+const ANIMATION_STYLES_MASK_FINAL = {'opacity': ''};
+
+/**
+ * @param {T} current
+ * @return {{current: T}}
+ * @template T
+ */
+function useValueRef(current) {
```
Comment (caroqliu) [inline extensions/amp-sidebar/1.0/sidebar.js:54]: I think lightbox also defines `useValueRef`, do you want to move it to `preact/index` or `preact/utils` instead?

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 9 above)
Comment (caroqliu) [reply extensions/amp-sidebar/1.0/sidebar.js:63]: Should `safelySetStyles` be moved to `style.js`?

File: extensions/amp-sidebar/1.0/sidebar.js [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+    children,
+    ...rest
+  },
+  ref
+) {
+  // There are two phases to open and close.
+  // To open, we mount and render the contents (invisible), then animate the display (visible).
+  // To close, it's the reverse.
+  // `mounted` mounts the component. `opened` plays the animation.
+  const initialOpen = false;
```
Comment (caroqliu) [inline extensions/amp-sidebar/1.0/sidebar.js:87]: If this is always `false`, do we need a named constant for the two times it's used?

File: extensions/amp-sidebar/1.0/sidebar.js [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+  const open = useCallback(() => {
+    if (onBeforeOpenRef.current) {
+      onBeforeOpenRef.current();
+    }
+    setMounted(true);
+    setOpened(true);
+  }, [onBeforeOpenRef]);
+  const close = useCallback(() => setOpened(false), []);
+  const toggle = useCallback(() => (opened ? close() : open()), [
+    opened,
```
Comment (caroqliu) [inline extensions/amp-sidebar/1.0/sidebar.js:109]: Would it be helpful to have an `openedRef` so this callback value does not have to be redefined every time the state changes?

File: extensions/amp-sidebar/1.0/sidebar.js [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+  },
+  ref
+) {
+  // There are two phases to open and close.
+  // To open, we mount and render the contents (invisible), then animate the display (visible).
+  // To close, it's the reverse.
+  // `mounted` mounts the component. `opened` plays the animation.
+  const initialOpen = false;
+  const [mounted, setMounted] = useState(initialOpen);
+  const [opened, setOpened] = useState(initialOpen);
```
Comment (caroqliu) [inline extensions/amp-sidebar/1.0/sidebar.js:89]: nit: Should this follow `[visible, setVisible]` like lightbox? the meaning of "open" could get a bit overloaded here

File: extensions/amp-sidebar/1.0/sidebar.js [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+          : ANIMATION_STYLES_SIDEBAR_RIGHT_INIT
+      );
+      safelySetStyles(maskElement, ANIMATION_STYLES_MASK_INIT);
+      sidebarAnimation = sidebarElement.animate(
+        side === 'left'
+          ? ANIMATION_KEYFRAMES_SLIDE_IN_LEFT
+          : ANIMATION_KEYFRAMES_SLIDE_IN_RIGHT,
+        {
+          duration: ANIMATION_DURATION,
+          fill: 'both',
```
Comment (caroqliu) [inline extensions/amp-sidebar/1.0/sidebar.js:158]: FMI: Why are some of these style objects given constant doc level names and others inlined here?

File: extensions/amp-sidebar/1.0/sidebar.js [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+          as={Comp}
+          ref={sidebarRef}
+          size={false}
+          layout={true}
+          paint={true}
+          className={`${classes.sidebarClass} ${
+            side === 'left' ? classes.left : classes.right
+          }`}
+          role="menu"
+          tabindex="-1"
```
Comment (caroqliu) [inline extensions/amp-sidebar/1.0/sidebar.js:224]: Isn't a `<div>` not focusable by default?

File: extensions/amp-sidebar/1.0/sidebar.jss.js [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {createUseStyles} from 'react-jss';
+
+const sidebarClass = {
+  position: 'fixed !important',
```
Comment (caroqliu) [inline extensions/amp-sidebar/1.0/sidebar.jss.js:20]: FMI: Why do we need these `!important`s?

File: extensions/amp-sidebar/1.0/storybook/Basic.js [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+ * @return {*}
+ */
+function SidebarWithActions(props) {
+  // TODO(#30447): replace imperative calls with "button" knobs when the
+  // Storybook 6.1 is released.
+  const ref = Preact.useRef();
+  return (
+    <>
+      <Sidebar ref={ref} {...props}>
+        <div style={{margin: 8}}>
```
Comment (caroqliu) [inline extensions/amp-sidebar/1.0/storybook/Basic.js:39]: Is this margin necessary?

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 17 above)
Comment (caroqliu) [reply extensions/amp-sidebar/1.0/sidebar.js:224]: Also, shouldn't it be focusable so it can listen for `Esc` to close?

**Review by caroqliu — APPROVED** [review]
Left a couple comments/questions but LGTM for the foundation. I'm curious what the plan is for:
- checking/warning if the component is a direct descendent of `<body>`
- responsive toolbar feature
- future functionality for nested menus that interop with sidebar
- bonus: if we were to add support for display locking, what would that look like? we could no longer conditionally render the markup

Would also be nice for the storybook to have a scrollable sample too.

File: extensions/amp-sidebar/1.0/storybook/Basic.js (hunk 19 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/storybook/Basic.js:39]: It's not needed per se, but I think it looks infinitely better with the margin. 😂

Are you opposed to it?

File: extensions/amp-sidebar/1.0/sidebar.jss.js (hunk 18 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.jss.js:20]: Mostly just so the user cannot over overwrite by accident (or on purpose).  The way the sidebar animates in and is positioned on the viewport is all critically dependent on `position: fixed`.

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 17 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.js:224]: The only ways to close that have been implemented so far are the imperative API and click on the mask.  I have an action item in the tracker: https://github.com/ampproject/amphtml/issues/31366 for handling focus, so I'll follow up at that point :).  This may be updated.

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 16 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.js:158]: It is a tiny bit wonky, but my thought process was -

`fill` and `direction` are not configurable as they are fundamentally needed for the way the animation works.
`duration` and `ease-in` can be configured to make the animation slower / faster, curvier, and it still works, so thought it made sense to make as a constant.  

I'm not married to it either way so let me know if you feel strongly about it.

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 12 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.js:54]: At this very initial stage, I'd like to keep it separate as I think a lot may change during the rest of the sidebar implementation.  But I've added an item to the tracker to https://github.com/ampproject/amphtml/issues/31366 to look at shared items between this and lightbox.  We have also talked about creating a common component that does the animations.

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 15 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.js:89]: I personally prefer opened since the sidebar docs all refer to the sidebar as being opened / closed. 🙂

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 13 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.js]: Sure updated!

File: extensions/amp-sidebar/1.0/sidebar.js (hunk 14 above)
Comment (krdwan) [reply extensions/amp-sidebar/1.0/sidebar.js:108]: Thanks, will address this as a follow up once we get further into the sidebar implementation.

File: extensions/amp-sidebar/1.0/storybook/Basic.js (hunk 19 above)
Comment (dvoytenko) [reply extensions/amp-sidebar/1.0/storybook/Basic.js:39]: This is storybook only, right?

Comment (krdwan) [reply extensions/amp-sidebar/1.0/storybook/Basic.js:39]: Right, this is just in the storybook.
